### ISSUE TITLE
[Core] Incorrect detection of cpus

### ISSUE THREAD

**Issue by shara19** [ray-project/ray#34846]
### What happened + What you expected to happen

If I'm understanding correctly, Ray detects number of CPUs using [this](method). However, based on the multiprocessing [docs](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.cpu_count), this is not the number of usable cpus, instead we should be using `len(os.sched_getaffinity(0))` to get the usable cpus

I spent all afternoon trying to understanding why Ray was detecting the wrong number of cpus and noticed that  multiprocessing.cpu_count() gives the number of all the cpus on a given node - this is problematic if the node is shared resulting in uninformative error. `len(os.sched_getaffinity(0))`  gives the right number of CPUs, I can make a PR to fix this if it seems okay

Currently, the work around is to set the number of cpus in init
`ray.init(num_cpus=len(os.sched_getaffinity(0)))`

### Versions / Dependencies

Slurm on HPC

### Reproduction script

I don't think it can be reproduced unless you have a system with shared resources and some are restricted

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by akshara08** [comment]
Yeah, makes sense. Looks like`sched_getaffinity` is only available on some UNIX systems and I'm assuming the cases where `multiprocessing.cpu_count()` is different from `len(os.sched_getaffinity(0))` are in systems that allow shared resources and hopefully expose those interfaces when `sched_getaffinity` is called. MacOS/Windows do not expose these interfaces causing the error, in those cases I think it's fair to fallback to `multiprocessing.cpu_count()`. 

I can make a PR in couple of days

**Comment by akshara08** [comment]
I'm sorry for the delay on this. Life happened. I should free up in the next couple of weeks. It's still on my radar

**Comment by anyscalesam** [comment]
@akshara08 did you get a chance to work on that PR? Can you link it here?

**Comment by Superskyyy** [comment]
I can take over if @akshara08 is away. @anyscalesam please assign me.

**Comment by Superskyyy** [comment]
`os.cpu_count()`  is the same as multiprocessing.cpu_count() so both will be replaced in the codebase based on usage.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
